File: algorithms/leetcode/226-invert-binary-tree/invert.cpp
// SPDX-License-Identifier: MIT
// Copyright (c) 2019 Gennady Trafimenkov

#include <iostream>
#include <string>
#include <sstream>

struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode(int x) : val(x), left(NULL), right(NULL) {}

    std::string GetString() {
        std::ostringstream ss;
        ss << '(' << this->val
            << ' ' << (this->left ? this->left->GetString() : "null")
            << ' ' << (this->right ? this->right->GetString() : "null")
            << ')';
        return ss.str();
    }

    void Insert(int x) {
        if (this->val < x) {
            this->right = UpdateLeaf(this->right, x);
        }
        if (this->val > x) {
            this->left = UpdateLeaf(this->left, x);
        }
    }

    ~TreeNode() {
        if (left) delete left;
        if (right) delete right;
    }

    private:

    static TreeNode* UpdateLeaf(TreeNode* leaf, int x) {
        if (!leaf) {
            return new TreeNode(x);
        }
        leaf->Insert(x);
        return leaf;
    }
};

class Solution {
public:
    TreeNode* invertTree(TreeNode* root) {
        if (root) {
            TreeNode* inverted = new TreeNode(root->val);
            inverted->left = invertTree(root->right);
            inverted->right = invertTree(root->left);
            return inverted;
        }
        return NULL;
    }
};

int main() {
    TreeNode* root = new TreeNode(4);
    root->Insert(2);
    root->Insert(7);
    root->Insert(1);
    root->Insert(3);
    root->Insert(6);
    root->Insert(9);
    std::cout << root->GetString() << std::endl;

    Solution sol;
    TreeNode* inverted = sol.invertTree(root);
    std::cout << inverted->GetString() << std::endl;

    delete inverted;
    delete root;

    return 0;
}

File: algorithms/leetcode/226-invert-binary-tree/invert.py
# SPDX-License-Identifier: MIT
# Copyright (c) 2019 Gennady Trafimenkov

class TreeNode:
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None

    def __str__(self):
        return "({} {} {})".format(self.val, self.left, self.right)

    def insert(self, x):
        def _updateLeaf(leaf, x):
            if leaf == None:
                return TreeNode(x)
            leaf.insert(x)
            return leaf
        if self.val < x:
            self.right = _updateLeaf(self.right, x)
        if self.val > x:
            self.left = _updateLeaf(self.left, x)

class Solution:
    def invertTree(self, root):
        """
        :type root: TreeNode
        :rtype: TreeNode
        """
        if root == None:
            return None
        new_root = TreeNode(root.val)
        new_root.left = self.invertTree(root.right)
        new_root.right = self.invertTree(root.left)
        return new_root

if __name__ == '__main__':
    root = TreeNode(4)
    root.insert(2)
    root.insert(7)
    root.insert(1)
    root.insert(3)
    root.insert(6)
    root.insert(9)
    print(root)
    sol = Solution()
    inverted = sol.invertTree(root)
    print(inverted)
